File: partyplug/RoomFlow.js
'use strict';

// RoomFlow owns room state, roster identity/join order, presence, and host
// election. It does not own DOM, transport, countdown timers, or game fields
// like color/name/score; those are opaque data on the live player record.

(function (root, factory) {
  var RoomFlow = factory();
  if (typeof module !== 'undefined' && module.exports) {
    module.exports = RoomFlow;
  } else {
    root.RoomFlow = RoomFlow;
  }
})(typeof self !== 'undefined' ? self : this, function () {

  var STATES = Object.freeze({
    LOBBY: 'lobby',
    COUNTDOWN: 'countdown',
    PLAYING: 'playing',
    RESULTS: 'results',
  });

  var VALID_TRANSITIONS = {};
  VALID_TRANSITIONS[STATES.LOBBY] = [STATES.COUNTDOWN];
  VALID_TRANSITIONS[STATES.COUNTDOWN] = [STATES.PLAYING, STATES.LOBBY];
  VALID_TRANSITIONS[STATES.PLAYING] = [STATES.RESULTS, STATES.LOBBY];
  VALID_TRANSITIONS[STATES.RESULTS] = [STATES.COUNTDOWN, STATES.LOBBY];

  function RoomFlow(opts) {
    opts = opts || {};
    // Optional () => peerIndex. When the transport designates a master
    // controller (AirConsole), supply it here. Returns null/undefined when
    // there is no platform master.
    this.masterProvider = typeof opts.masterProvider === 'function' ? opts.masterProvider : null;

    this.state = STATES.LOBBY;
    this.players = new Map();        // peerIndex -> player record
    this.hostPeerIndex = null;       // sticky host slot (raw; see `host` getter for effective)
    this._joinSeq = 0;               // monotonic joinedAt source (Date.now collides in same ms)
    this._disconnected = new Set();  // peerIndices currently in the disconnect window
    this._order = [];                // active participants (snapshotted on COUNTDOWN, or via setActiveOrder)
    this._listeners = {};
  }

  RoomFlow.STATES = STATES;

  // Lowest free dense slot in [0, max). Pass slot values, not peerIndices, so
  // sparse transport ids (e.g. AirConsole device_id) don't become color/seat ids.
  RoomFlow.lowestFreeSlot = function (used, max) {
    var taken = used instanceof Set ? used : new Set(used);
    for (var i = 0; i < max; i++) { if (!taken.has(i)) return i; }
    return -1;
  };

  // ---- tiny event emitter (dependency-free; portable Node + browser) ----
  RoomFlow.prototype.on = function (type, handler) {
    (this._listeners[type] = this._listeners[type] || []).push(handler);
    var self = this;
    return function () { self.off(type, handler); };
  };
  RoomFlow.prototype.off = function (type, handler) {
    var arr = this._listeners[type];
    if (!arr) return;
    var i = arr.indexOf(handler);
    if (i >= 0) arr.splice(i, 1);
  };
  RoomFlow.prototype._emit = function (type, detail) {
    var arr = this._listeners[type];
    if (arr) { var copy = arr.slice(); for (var i = 0; i < copy.length; i++) copy[i](detail); }
    var wild = this._listeners['*'];
    if (wild) { var w = wild.slice(); for (var j = 0; j < w.length; j++) w[j](type, detail); }
  };

  // =====================================================================
  // Roster
  // =====================================================================

  // Add a player, or reconnect/refresh an existing one (same peerIndex).
  // `fields` is opaque game data merged onto the record (name, color slot,
  // level, ...). RoomFlow adds peerIndex/joinedAt/connected and never reads
  // the game fields. Returns the live player record.
  RoomFlow.prototype.addPlayer = function (peerIndex, fields) {
    fields = fields || {};
    var existing = this.players.get(peerIndex);
    if (existing) {
      var prevHost = this.host;
      // Reconnect: keep slot / joinedAt / host, refresh presence + fields.
      // Protect kit-owned fields against a caller passing them in `fields`
      // (e.g. a serialized record round-tripped through JSON): joinedAt is the
      // host-election tiebreak and peerIndex is the map key — neither may be
      // clobbered by game data.
      var savedJoinedAt = existing.joinedAt;
      Object.assign(existing, fields);
      existing.joinedAt = savedJoinedAt;
      existing.peerIndex = peerIndex;
      existing.connected = true;
      this._disconnected.delete(peerIndex);
      if (this.host !== prevHost) this._emit('hostchange', { hostPeerIndex: this.host });
      this._emit('playerupdate', { player: existing });
      this._emit('rosterchange', { players: this.list() });
      return existing;
    }
    var player = Object.assign({}, fields, {
      peerIndex: peerIndex,
      joinedAt: this._joinSeq++,
      connected: true,
    });
    this.players.set(peerIndex, player);
    // First joiner owns the sticky host slot. Also covers the "room emptied
    // then someone joined" case (hostPeerIndex was reset to null).
    if (this.hostPeerIndex == null) {
      this.hostPeerIndex = peerIndex;
      this._emit('hostchange', { hostPeerIndex: this.host });
    }
    this._emit('playerjoin', { player: player });
    this._emit('rosterchange', { players: this.list() });
    return player;
  };

  // Hard leave (peer_left). The sticky slot only moves when the holder
  // departs from LOBBY/RESULTS; a mid-game leave leaves the slot untouched
  // so a reconnecting host reclaims it (the `host` getter falls back
  // meanwhile). hostchange fires whenever the EFFECTIVE host changes — including
  // a mid-game departure where the sticky slot stays put but the getter's
  // fallback shifts to a present player.
  RoomFlow.prototype.removePlayer = function (peerIndex) {
    if (!this.players.has(peerIndex)) return;
    var prevHost = this.host;
    var wasHost = peerIndex === this.hostPeerIndex;
    this.players.delete(peerIndex);
    this._disconnected.delete(peerIndex);
    var oi = this._order.indexOf(peerIndex);
    if (oi >= 0) this._order.splice(oi, 1);
    if (wasHost && (this.state === STATES.LOBBY || this.state === STATES.RESULTS)) {
      this.hostPeerIndex = this._electNextHost(peerIndex);
    }
    if (this.host !== prevHost) this._emit('hostchange', { hostPeerIndex: this.host });
    this._emit('playerleave', { peerIndex: peerIndex });
    this._emit('rosterchange', { players: this.list() });
  };

  // Re-key a player from one peerIndex to another. This is ONLY for cross-device
  // takeover: a different client (fresh clientId) claims a dropped player's
  // still-present slot and gets a new peerIndex from the relay. A same-client
  // reconnect keeps its index (the relay keys slots by clientId) and never needs
  // this. Preserves the record (incl. joinedAt) and rekeys host slot + order.
  RoomFlow.prototype.rekey = function (oldId, newId) {
    if (oldId === newId) return false;
    var rec = this.players.get(oldId);
    if (!rec) return false;
    var prevHost = this.host;
    this.players.delete(oldId);
    this.players.delete(newId); // drop the placeholder slot the returning peer got
    rec.peerIndex = newId;
    rec.connected = true;
    this.players.set(newId, rec);
    this._disconnected.delete(oldId);
    this._disconnected.delete(newId);
    for (var i = 0; i < this._order.length; i++) {
      if (this._order[i] === oldId) this._order[i] = newId;
    }
    // The slot wasn't moved when this player blipped mid-game, so if it still
    // points at the old peerIndex, rekey it so the reconnecting host resumes.
    // Only the host's own slot is rekeyed; a non-host claim never promotes
    // (when there's no sticky host, the `host` getter's oldest-eligible
    // fallback already picks the right player).
    if (this.hostPeerIndex === oldId) {
      this.hostPeerIndex = newId;
    }
    if (this.host !== prevHost) this._emit('hostchange', { hostPeerIndex: this.host });
    this._emit('rosterchange', { players: this.list() });
    return true;
  };

  // Soft disconnect window (the player record stays; presence flips false).
  // Emits hostchange if the effective host shifts (e.g. the host blips mid-game
  // and the getter's fallback hands duty to a present player).
  RoomFlow.prototype.markDisconnected = function (peerIndex) {
    var p = this.players.get(peerIndex);
    if (!p) return;
    var prevHost = this.host;
    p.connected = false;
    this._disconnected.add(peerIndex);
    if (this.host !== prevHost) this._emit('hostchange', { hostPeerIndex: this.host });
    this._emit('rosterchange', { players: this.list() });
  };

  RoomFlow.prototype.markReconnected = function (peerIndex) {
    var p = this.players.get(peerIndex);
    if (!p) return;
    var prevHost = this.host;
    p.connected = true;
    this._disconnected.delete(peerIndex);
    if (this.host !== prevHost) this._emit('hostchange', { hostPeerIndex: this.host });
    this._emit('rosterchange', { players: this.list() });
  };

  // Clear every disconnect flag, marking all current players present. Used at
  // game start / lobby return where stale blip flags must not suppress host
  // eligibility for the new round.
  RoomFlow.prototype.clearDisconnected = function () {
    if (this._disconnected.size === 0) return;
    var prevHost = this.host;
    this._disconnected.clear();
    for (var entry of this.players) entry[1].connected = true;
    if (this.host !== prevHost) this._emit('hostchange', { hostPeerIndex: this.host });
    this._emit('rosterchange', { players: this.list() });
  };

  // =====================================================================
  // Host election (DisplayState.getHostPeerIndex / electNextHost / reconcile)
  // =====================================================================

  // During COUNTDOWN/PLAYING/RESULTS the candidate set is restricted to the
  // active participants (the `_order`), so a late joiner can't be handed host
  // duty for menu actions they can't reach. Open to everyone in LOBBY.
  RoomFlow.prototype._restricted = function () {
    return (this.state === STATES.COUNTDOWN ||
            this.state === STATES.PLAYING ||
            this.state === STATES.RESULTS) && this._order.length > 0;
  };

  RoomFlow.prototype._isEligible = function (peerIndex, eligibleSet) {
    return peerIndex != null &&
      this.players.has(peerIndex) &&
      !this._disconnected.has(peerIndex) &&
      (eligibleSet == null || eligibleSet.has(peerIndex));
  };

  // Oldest-joined present player within eligibleSet (null = everyone present),
  // optionally skipping excludeId. Shared by the `host` getter (no exclusion)
  // and _electNextHost (excludes the departing holder).
  RoomFlow.prototype._oldestEligible = function (eligibleSet, excludeId) {
    var bestId = null, bestJoin = Infinity;
    for (var entry of this.players) {
      var id = entry[0];
      if (id === excludeId) continue;
      if (this._disconnected.has(id)) continue;
      if (eligibleSet != null && !eligibleSet.has(id)) continue;
      var ja = entry[1].joinedAt == null ? Infinity : entry[1].joinedAt;
      if (ja < bestJoin) { bestJoin = ja; bestId = id; }
    }
    return bestId;
  };

  // Effective host: platform master (if eligible) -> sticky host (if
  // eligible) -> oldest-joined eligible present player. Read-only; the
  // sticky slot is only mutated by removePlayer / rekey / reconcile.
  Object.defineProperty(RoomFlow.prototype, 'host', {
    get: function () {
      var restricted = this._restricted();
      var eligible = restricted ? new Set(this._order) : null;
      if (this.masterProvider) {
        var m = this.masterProvider();
        if (this._isEligible(m, eligible)) return m;
      }
      if (this._isEligible(this.hostPeerIndex, eligible)) return this.hostPeerIndex;
      return this._oldestEligible(eligible);
    },
  });

  RoomFlow.prototype.isHost = function (peerIndex) {
    return peerIndex != null && peerIndex === this.host;
  };

  // Oldest-joined present player other than excludeId. Restricted to the
  // participant order while in COUNTDOWN/PLAYING/RESULTS, so committing the
  // sticky slot can never promote a late joiner (who is not in `_order`) over
  // the actual game participants. Unrestricted in LOBBY, where everyone present
  // is a valid candidate. Returns null when nobody qualifies.
  RoomFlow.prototype._electNextHost = function (excludeId) {
    return this._oldestEligible(this._restricted() ? new Set(this._order) : null, excludeId);
  };

  // Commit any pending sticky-host handoff. Called when entering LOBBY or
  // RESULTS (the moments host duty is actually exercised: Start, Play Again).
  RoomFlow.prototype._reconcileStickyHost = function () {
    if (this.players.size === 0) return;
    var eligible = this._restricted() ? new Set(this._order) : null;
    if (this.hostPeerIndex != null &&
        this.players.has(this.hostPeerIndex) &&
        !this._disconnected.has(this.hostPeerIndex) &&
        (eligible == null || eligible.has(this.hostPeerIndex))) {
      return;
    }
    var prev = this.hostPeerIndex;
    this.hostPeerIndex = this._electNextHost(this.hostPeerIndex);
    if (this.hostPeerIndex !== prev) this._emit('hostchange', { hostPeerIndex: this.host });
  };

  // =====================================================================
  // Lifecycle
  // =====================================================================

  // Snapshot the current connected roster as the active participant order
  // (join order). Called automatically when entering COUNTDOWN.
  RoomFlow.prototype._snapshotOrder = function () {
    var active = [];
    for (var entry of this.players) {
      if (!this._disconnected.has(entry[0])) active.push(entry[1]);
    }
    active.sort(function (a, b) { return a.joinedAt - b.joinedAt; });
    this._order = active.map(function (p) { return p.peerIndex; });
  };

  // Let a game that maintains its own participant order (e.g. for board
  // layout) keep RoomFlow's host-eligibility set exactly in sync with it.
  RoomFlow.prototype.setActiveOrder = function (peerIndices) {
    var out = [];
    for (var i = 0; i < (peerIndices || []).length; i++) {
      if (this.players.has(peerIndices[i])) out.push(peerIndices[i]);
    }
    this._order = out;
  };

  // Validated state transition. Public so games that run their own countdown
  // (like HexStacker) can drive the machine imperatively; the high-level
  // helpers below call it too. Returns true if applied.
  RoomFlow.prototype.transitionTo = function (to) {
    var from = this.state;
    if (to === from) return true;
    var allowed = VALID_TRANSITIONS[from];
    if (!allowed || allowed.indexOf(to) < 0) {
      if (typeof console !== 'undefined' && console.warn) {
        console.warn('RoomFlow: invalid transition ' + from + ' -> ' + to);
      }
      return false;
    }
    this.state = to;
    if (to === STATES.COUNTDOWN) this._snapshotOrder();
    if (to === STATES.LOBBY) this._order = [];
    if (to === STATES.LOBBY || to === STATES.RESULTS) this._reconcileStickyHost();
    this._emit('statechange', { from: from, to: to });
    return true;
  };

  // Lifecycle helpers — thin, validated transitions. The countdown itself is
  // game-owned (its visuals and controller messaging are game-flavored): a game
  // drives transitionTo(COUNTDOWN) -> run its own countdown -> transitionTo(PLAYING).

  // Readable sugar for `transitionTo('results')`. Results data is the game's
  // own (it knows the scoring); the kit does not store it.
  RoomFlow.prototype.endGame = function () {
    return this.transitionTo(STATES.RESULTS);
  };

  RoomFlow.prototype.returnToLobby = function () {
    return this.transitionTo(STATES.LOBBY);
  };

  // =====================================================================
  // Read accessors
  // =====================================================================

  // Roster as an array sorted by join order.
  RoomFlow.prototype.list = function () {
    var arr = [];
    for (var entry of this.players) arr.push(entry[1]);
    arr.sort(function (a, b) { return a.joinedAt - b.joinedAt; });
    return arr;
  };

  RoomFlow.prototype.get = function (peerIndex) { return this.players.get(peerIndex) || null; };
  RoomFlow.prototype.has = function (peerIndex) { return this.players.has(peerIndex); };

  Object.defineProperty(RoomFlow.prototype, 'size', {
    get: function () { return this.players.size; },
  });

  // Connected player count (what a lobby "Start (N)" button should show).
  Object.defineProperty(RoomFlow.prototype, 'connectedCount', {
    get: function () {
      var n = 0;
      for (var entry of this.players) { if (!this._disconnected.has(entry[0])) n++; }
      return n;
    },
  });

  RoomFlow.prototype.isDisconnected = function (peerIndex) { return this._disconnected.has(peerIndex); };

  // Reset to a fresh room (new room / return to welcome). Mirrors the
  // roster/host/state portion of DisplayState.resetRoomData.
  RoomFlow.prototype.reset = function () {
    var prevState = this.state;
    var hadHost = this.hostPeerIndex != null;
    // IMPORTANT: clear the Map in place — never reassign `this.players`.
    // Consumers may alias this exact Map object as their roster (HexStacker's
    // DisplayState does), so reassigning would leave them on a stale Map.
    this.players.clear();
    this._disconnected.clear();
    this._order = [];
    this.hostPeerIndex = null;
    this._joinSeq = 0;
    this.state = STATES.LOBBY;
    // Emit so event-driven consumers re-render on reset — every other state
    // change goes through transitionTo (which emits), so reset must too or a
    // game subscribed to statechange/rosterchange would miss the room wipe.
    if (prevState !== STATES.LOBBY) this._emit('statechange', { from: prevState, to: STATES.LOBBY });
    this._emit('rosterchange', { players: [] });
    if (hadHost) this._emit('hostchange', { hostPeerIndex: null });
  };

  return RoomFlow;
});

File: native/libttp-party/ttp/room_flow.cc
#include "ttp/room_flow.h"

#include <algorithm>

namespace ttp {

// ---- Player serialization ---------------------------------------------------
Value Player::toValue() const {
  Value o = Value::Obj();
  for (const auto& kv : fields) o.set(kv.first, kv.second);
  o.set("peerIndex", peerIndex.toValue());
  o.set("joinedAt", Value::Num(joinedAt));
  o.set("connected", Value::Bool(connected));
  return o;
}

// ---- valid transitions ------------------------------------------------------
// LOBBY -> [COUNTDOWN]; COUNTDOWN -> [PLAYING, LOBBY];
// PLAYING -> [RESULTS, LOBBY]; RESULTS -> [COUNTDOWN, LOBBY].
static bool transitionAllowed(const std::string& from, const std::string& to) {
  if (from == "lobby") return to == "countdown";
  if (from == "countdown") return to == "playing" || to == "lobby";
  if (from == "playing") return to == "results" || to == "lobby";
  if (from == "results") return to == "countdown" || to == "lobby";
  return false;
}

// ============================================================================
RoomFlow::RoomFlow(const Config& cfg, EmitFn emit) : cfg_(cfg), emit_(std::move(emit)) {
  masterValue_ = cfg_.master;
}

bool RoomFlow::isKitKey(const std::string& k) {
  return k == "peerIndex" || k == "joinedAt" || k == "connected";
}

std::string RoomFlow::stateName() const {
  switch (state_) {
    case State::LOBBY: return "lobby";
    case State::COUNTDOWN: return "countdown";
    case State::PLAYING: return "playing";
    case State::RESULTS: return "results";
  }
  return "lobby";
}

// ---- roster storage helpers (JS Map semantics over an ordered vector) -------
Player* RoomFlow::find(const PeerId& id) {
  for (auto& p : players_) if (p.peerIndex == id) return &p;
  return nullptr;
}
const Player* RoomFlow::find(const PeerId& id) const {
  for (const auto& p : players_) if (p.peerIndex == id) return &p;
  return nullptr;
}
void RoomFlow::erasePlayer(const PeerId& id) {
  for (size_t i = 0; i < players_.size(); i++)
    if (players_[i].peerIndex == id) { players_.erase(players_.begin() + i); return; }
}

bool RoomFlow::discHas(const PeerId& id) const {
  for (const auto& d : disconnected_) if (d == id) return true;
  return false;
}
void RoomFlow::discAdd(const PeerId& id) {
  if (!discHas(id)) disconnected_.push_back(id);
}
void RoomFlow::discDelete(const PeerId& id) {
  for (size_t i = 0; i < disconnected_.size(); i++)
    if (disconnected_[i] == id) { disconnected_.erase(disconnected_.begin() + i); return; }
}

bool RoomFlow::lastSeenHas(const PeerId& id) const {
  for (const auto& kv : lastSeen_) if (kv.first == id) return true;
  return false;
}
bool RoomFlow::lastSeenGet(const PeerId& id, double& out) const {
  for (const auto& kv : lastSeen_) if (kv.first == id) { out = kv.second; return true; }
  return false;
}
void RoomFlow::lastSeenSet(const PeerId& id, double v) {
  for (auto& kv : lastSeen_) if (kv.first == id) { kv.second = v; return; }
  lastSeen_.emplace_back(id, v);
}
void RoomFlow::lastSeenDelete(const PeerId& id) {
  for (size_t i = 0; i < lastSeen_.size(); i++)
    if (lastSeen_[i].first == id) { lastSeen_.erase(lastSeen_.begin() + i); return; }
}

// ---- emit helpers -----------------------------------------------------------
void RoomFlow::emitHostchange() {
  Value d = Value::Obj();
  d.set("hostPeerIndex", host().toValue());
  emit("hostchange", std::move(d));
}
void RoomFlow::emitRosterchange() {
  Value d = Value::Obj();
  d.set("players", listValue());
  emit("rosterchange", std::move(d));
}

// ============================================================================
// Roster
// ============================================================================
const Player* RoomFlow::addPlayer(const PeerId& peerIndex,
                                  const std::vector<std::pair<std::string, Value>>& fields) {
  Player* existing = find(peerIndex);
  if (existing) {
    PeerId prevHost = host();
    // Reconnect: merge game fields (Object.assign), protect the kit-owned keys
    // (joinedAt is the election tiebreak, peerIndex is the map key), refresh
    // presence. joinedAt/peerIndex stay; connected -> true.
    for (const auto& kv : fields) {
      if (isKitKey(kv.first)) continue;
      bool set = false;
      for (auto& f : existing->fields) if (f.first == kv.first) { f.second = kv.second; set = true; break; }
      if (!set) existing->fields.emplace_back(kv.first, kv.second);
    }
    existing->connected = true;
    discDelete(peerIndex);
    if (host() != prevHost) emitHostchange();
    Value d = Value::Obj();
    d.set("player", existing->toValue());
    emit("playerupdate", std::move(d));
    emitRosterchange();
    return existing;
  }
  Player p;
  p.peerIndex = peerIndex;
  p.joinedAt = joinSeq_++;
  p.connected = true;
  for (const auto& kv : fields) if (!isKitKey(kv.first)) p.fields.emplace_back(kv.first, kv.second);
  players_.push_back(std::move(p));
  Player* added = &players_.back();
  // First joiner owns the sticky host slot (also the emptied-then-rejoined case).
  if (hostPeerIndex_.isNull()) {
    hostPeerIndex_ = peerIndex;
    emitHostchange();
  }
  Value d = Value::Obj();
  d.set("player", added->toValue());
  emit("playerjoin", std::move(d));
  emitRosterchange();
  return added;
}

void RoomFlow::removePlayer(const PeerId& peerIndex) {
  if (!find(peerIndex)) return;
  PeerId prevHost = host();
  bool wasHost = peerIndex == hostPeerIndex_;
  erasePlayer(peerIndex);
  discDelete(peerIndex);
  lastSeenDelete(peerIndex);
  for (size_t i = 0; i < order_.size(); i++)
    if (order_[i] == peerIndex) { order_.erase(order_.begin() + i); break; }
  if (wasHost && (state_ == State::LOBBY || state_ == State::RESULTS)) {
    hostPeerIndex_ = electNextHost(&peerIndex);
  }
  if (host() != prevHost) emitHostchange();
  Value d = Value::Obj();
  d.set("peerIndex", peerIndex.toValue());
  emit("playerleave", std::move(d));
  emitRosterchange();
}

bool RoomFlow::rekey(const PeerId& oldId, const PeerId& newId) {
  if (oldId == newId) return false;
  Player* rec = find(oldId);
  if (!rec) return false;
  PeerId prevHost = host();
  Player moved = *rec;               // capture before the deletes
  erasePlayer(oldId);
  erasePlayer(newId);                // drop the placeholder slot the returning peer got
  moved.peerIndex = newId;
  moved.connected = true;
  players_.push_back(std::move(moved));  // re-appended -> moves to end of Map order
  discDelete(oldId);
  discDelete(newId);
  // Keep the NEWER of the two last-seen stamps (the placeholder's is the live
  // signal; the old seat's may be a full grace window stale).
  double oldStamp = 0, newStamp = 0;
  bool hasOld = lastSeenGet(oldId, oldStamp), hasNew = lastSeenGet(newId, newStamp);
  lastSeenDelete(oldId);
  lastSeenDelete(newId);
  if (hasOld || hasNew) {
    double kept = !hasOld ? newStamp : (!hasNew ? oldStamp : std::max(oldStamp, newStamp));
    lastSeenSet(newId, kept);
  }
  for (auto& id : order_) if (id == oldId) id = newId;
  if (hostPeerIndex_ == oldId) hostPeerIndex_ = newId;
  if (host() != prevHost) emitHostchange();
  emitRosterchange();
  return true;
}

void RoomFlow::markDisconnected(const PeerId& peerIndex) {
  Player* p = find(peerIndex);
  if (!p) return;
  PeerId prevHost = host();
  p->connected = false;
  discAdd(peerIndex);
  if (host() != prevHost) emitHostchange();
  emitRosterchange();
}

void RoomFlow::markReconnected(const PeerId& peerIndex) {
  Player* p = find(peerIndex);
  if (!p) return;
  PeerId prevHost = host();
  p->connected = true;
  discDelete(peerIndex);
  if (host() != prevHost) emitHostchange();
  emitRosterchange();
}

void RoomFlow::clearDisconnected(bool hasNow, double nowMs) {
  // Re-stamp liveness on this "everyone present" transition BEFORE the early
  // return, so a controller that just went quiet isn't instantly re-expired.
  if (hasNow) {
    for (const auto& p : players_) lastSeenSet(p.peerIndex, nowMs);
  }
  if (disconnected_.empty()) return;
  PeerId prevHost = host();
  disconnected_.clear();
  for (auto& p : players_) p.connected = true;
  if (host() != prevHost) emitHostchange();
  emitRosterchange();
}

// ============================================================================
// Host election
// ============================================================================
bool RoomFlow::restricted() const {
  return (state_ == State::COUNTDOWN || state_ == State::PLAYING || state_ == State::RESULTS) &&
         !order_.empty();
}

bool RoomFlow::isEligible(const PeerId& id, const std::vector<PeerId>* eligible) const {
  if (id.isNull()) return false;
  if (!find(id)) return false;
  if (discHas(id)) return false;
  if (eligible) {
    bool in = false;
    for (const auto& e : *eligible) if (e == id) { in = true; break; }
    if (!in) return false;
  }
  return true;
}

PeerId RoomFlow::oldestEligible(const std::vector<PeerId>* eligible, const PeerId* exclude) const {
  PeerId best = PeerId::None();
  double bestJoin = INFINITY;
  for (const auto& p : players_) {  // insertion order; joinedAt is unique so ties never arise
    if (exclude && p.peerIndex == *exclude) continue;
    if (discHas(p.peerIndex)) continue;
    if (eligible) {
      bool in = false;
      for (const auto& e : *eligible) if (e == p.peerIndex) { in = true; break; }
      if (!in) continue;
    }
    if (p.joinedAt < bestJoin) { bestJoin = p.joinedAt; best = p.peerIndex; }
  }
  return best;
}

PeerId RoomFlow::host() const {
  bool r = restricted();
  std::vector<PeerId> orderSet;
  const std::vector<PeerId>* eligible = nullptr;
  if (r) { orderSet = order_; eligible = &orderSet; }
  if (cfg_.hasMasterProvider) {
    PeerId m = masterValue_;
    if (isEligible(m, eligible)) return m;
  }
  if (isEligible(hostPeerIndex_, eligible)) return hostPeerIndex_;
  return oldestEligible(eligible, nullptr);
}

bool RoomFlow::isHost(const PeerId& peerIndex) const {
  return !peerIndex.isNull() && peerIndex == host();
}

PeerId RoomFlow::electNextHost(const PeerId* exclude) const {
  std::vector<PeerId> orderSet;
  const std::vector<PeerId>* eligible = nullptr;
  if (restricted()) { orderSet = order_; eligible = &orderSet; }
  return oldestEligible(eligible, exclude);
}

void RoomFlow::reconcileStickyHost() {
  if (players_.empty()) return;
  std::vector<PeerId> orderSet;
  const std::vector<PeerId>* eligible = nullptr;
  if (restricted()) { orderSet = order_; eligible = &orderSet; }
  if (!hostPeerIndex_.isNull() && find(hostPeerIndex_) && !discHas(hostPeerIndex_)) {
    bool ok = true;
    if (eligible) {
      ok = false;
      for (const auto& e : *eligible) if (e == hostPeerIndex_) { ok = true; break; }
    }
    if (ok) return;
  }
  PeerId prev = hostPeerIndex_;
  hostPeerIndex_ = electNextHost(&hostPeerIndex_);
  if (hostPeerIndex_ != prev) emitHostchange();
}

// ============================================================================
// Lifecycle
// ============================================================================
void RoomFlow::snapshotOrder() {
  std::vector<const Player*> active;
  for (const auto& p : players_) if (!discHas(p.peerIndex)) active.push_back(&p);
  std::stable_sort(active.begin(), active.end(),
                   [](const Player* a, const Player* b) { return a->joinedAt < b->joinedAt; });
  order_.clear();
  for (const Player* p : active) order_.push_back(p->peerIndex);
}

void RoomFlow::setActiveOrder(const std::vector<PeerId>& peerIndices) {
  std::vector<PeerId> out;
  for (const auto& id : peerIndices) if (find(id)) out.push_back(id);
  order_ = std::move(out);
}

bool RoomFlow::transitionTo(const std::string& to) {
  std::string from = stateName();
  if (to == from) return true;
  if (!transitionAllowed(from, to)) return false;  // console.warn is side-channel only
  if (to == "lobby") state_ = State::LOBBY;
  else if (to == "countdown") state_ = State::COUNTDOWN;
  else if (to == "playing") state_ = State::PLAYING;
  else if (to == "results") state_ = State::RESULTS;
  if (to != "playing") { hasGraceDeadline_ = false; }
  if (to == "countdown") snapshotOrder();
  if (to == "lobby") order_.clear();
  if (to == "lobby" || to == "results") reconcileStickyHost();
  Value d = Value::Obj();
  d.set("from", Value::Str(from));
  d.set("to", Value::Str(to));
  emit("statechange", std::move(d));
  return true;
}

// ============================================================================
// Liveness (pure, nowMs-injected predicates)
// ============================================================================
void RoomFlow::onSeen(const PeerId& peerIndex, double nowMs) {
  if (find(peerIndex)) lastSeenSet(peerIndex, nowMs);
}

bool RoomFlow::isExpired(const PeerId& peerIndex, double nowMs) const {
  if (cfg_.hasEnabledProvider && !livenessEnabled_) return false;
  double stamp = 0;
  if (!lastSeenGet(peerIndex, stamp)) return false;
  return nowMs - stamp > cfg_.timeoutMs;  // strict '>' — exactly-at-timeout is alive
}

std::vector<PeerId> RoomFlow::expiredPeers(double nowMs) const {
  std::vector<PeerId> out;
  if (state_ == State::LOBBY) return out;
  for (const auto& p : players_) {  // Map key order -> return array order
    if (discHas(p.peerIndex)) continue;
    if (isExpired(p.peerIndex, nowMs)) out.push_back(p.peerIndex);
  }
  return out;
}

bool RoomFlow::allParticipantsDisconnected() const {
  if (order_.empty()) return false;
  for (const auto& id : order_) if (!discHas(id)) return false;
  return true;
}

bool RoomFlow::hasLateJoiners() const {
  for (const auto& p : players_) {
    bool inOrder = false;
    for (const auto& id : order_) if (id == p.peerIndex) { inOrder = true; break; }
    if (!inOrder) return true;
  }
  return false;
}

bool RoomFlow::graceTick(double nowMs) {
  if (state_ == State::PLAYING && allParticipantsDisconnected() && hasLateJoiners()) {
    if (!hasGraceDeadline_) { hasGraceDeadline_ = true; graceDeadline_ = nowMs + cfg_.graceMs; return false; }
    if (nowMs >= graceDeadline_) { hasGraceDeadline_ = false; return true; }
    return false;
  }
  hasGraceDeadline_ = false;
  return false;
}

// ============================================================================
// Read accessors
// ============================================================================
int RoomFlow::connectedCount() const {
  int n = 0;
  for (const auto& p : players_) if (!discHas(p.peerIndex)) n++;
  return n;
}

Value RoomFlow::listValue() const {
  std::vector<const Player*> arr;
  for (const auto& p : players_) arr.push_back(&p);
  std::stable_sort(arr.begin(), arr.end(),
                   [](const Player* a, const Player* b) { return a->joinedAt < b->joinedAt; });
  Value out = Value::Arr();
  for (const Player* p : arr) out.push(p->toValue());
  return out;
}

void RoomFlow::reset() {
  std::string prevState = stateName();
  bool hadHost = !hostPeerIndex_.isNull();
  players_.clear();
  disconnected_.clear();
  lastSeen_.clear();
  hasGraceDeadline_ = false;
  order_.clear();
  hostPeerIndex_ = PeerId::None();
  joinSeq_ = 0;
  state_ = State::LOBBY;
  if (prevState != "lobby") {
    Value d = Value::Obj();
    d.set("from", Value::Str(prevState));
    d.set("to", Value::Str("lobby"));
    emit("statechange", std::move(d));
  }
  {
    Value d = Value::Obj();
    d.set("players", Value::Arr());
    emit("rosterchange", std::move(d));
  }
  if (hadHost) {
    Value d = Value::Obj();
    d.set("hostPeerIndex", Value::Null());  // literal null, not host()
    emit("hostchange", std::move(d));
  }
}

}  // namespace ttp
